src/y0/algorithm/identify/id_star.py

```diff
@@ -0,0 +1,233 @@
+
+    """
+    # Rather than all n choose 2 combinations, we can restrict ourselves to the original graph variables and their counterfactual versions
+    return has_same_function(node1, node2) and has_same_parents(pw_graph, node1, node2)
+
+
+def lemma_25(
+    graph: NxMixedGraph[Variable], node1: Variable, node2: Variable
+) -> NxMixedGraph[Variable]:
+    r"""Merge node1 and node2 and return the reduced graph and query
+
+    Let :math:`M_\mathbf{x}` be a submodel derived from :math:`M` with set :math:`\mathbf{Z}` obse to attain values :math:`\mathbf{z}`, such that Lemma 24 holds for :math:`\alpha`; :math:`\beta`. Let :math:`M'` be a causal model obtained from :math:`M` by merging :math:`\alpha`; :math:`\beta` into a new node :math:`\omega`, which inherits all parents and the functional mechanism of :math:`\alpha`. All children of :math:`\alpha`; :math:`\beta` in :math:`M'` become children of :math:`\omega`. Then :math:`M_\mathbf{x},  M'_\mathbf{x} agree on any distribution consistent with :math:`z` being observed.
+
+    """
+    directed = [(u, v) for u, v in graph.directed.edges() if node2 not in (u, v)]
+    directed += [(node1, v) for u, v in graph.directed.edges() if node2 == u]
+    # directed += [(u, node1) for u, v in graph.directed.edges() if node2 == v]
+    undirected = [frozenset({u, v}) for u, v in graph.undirected.edges() if node2 not in (u, v)]
+    undirected += [
+        frozenset({node1, v}) for u, v in graph.undirected.edges() if node2 == u and node1 != v
+    ]
+    undirected += [
+        frozenset({u, node1}) for u, v in graph.undirected.edges() if node2 == v and node1 != u
+    ]
+    return NxMixedGraph.from_edges(
+        nodes=[node for node in graph.nodes() if node != node2],
+        directed=list(set(directed)),
+        undirected=[(u, v) for u, v in set(undirected)],
+    )
+
+
+def make_counterfactual_graph(
+    graph: NxMixedGraph[Variable], query: Probability
+) -> Tuple[NxMixedGraph[Variable], Probability]:
+    """Make counterfactual graph"""
+    worlds = get_worlds(query)
+    pw_graph = make_parallel_worlds_graph(graph, worlds)
+    new_query_variables = {v for v in query.get_variables() if not isinstance(v, Intervention)}
+    cf_graph = NxMixedGraph.from_edges(
+        nodes=pw_graph.nodes(),
+        directed=pw_graph.directed.edges(),
+        undirected=pw_graph.undirected.edges(),
+    )
+    for node in graph.topological_sort():
+        for intervention in worlds:
+            if (
```

Comment:
make this conditional an auxiliary function and give some docs as to what's getting checked